es_loader/es_loader.py

```diff
@@ -56,14 +124,27 @@ def _index_item(self, item):
 
         self.items_buffer.append(index_action)
 
-        if len(self.items_buffer) >= 500:
+        if len(self.items_buffer) >= self.buffer_size:
+            self.loaded_items_count += len(self.items_buffer)
             self._send_items()
             self.items_buffer = []
 
+        if len(self.items_buffer) < self.buffer_size \
+                and (self.items_count - self.loaded_items_count) < self.buffer_size and self.buffer_size > 0:
+            self.buffer_size = self.items_count - self.loaded_items_count
+            self.loaded_items_count += len(self.items_buffer)
+            if len(self.items_buffer) > 0:
+                self._send_items()
```

Comment:
In general, I see two problems repeating in this PR: over-splitting (if I can say so) of the code into tiny methods and overusing `self`.

I think the logic that is implemented here belongs to the outer loop, where it can be handled in a more natural way, approximately like this:

```
    def process_items(self):
        """Entry point to the whole process.
        Calls all methods in order to process items of the job.
        """
        batch = []
        for item in self._get_items():

            index_action = {
                '_op_type': 'index',
                '_index': self.index_name,
                '_type': self.doc_type,
                'doc': dict(item)
            }

            batch.append(index_action)

            # if we accumulated enough items, send them in a bulk:
            if len(batch) >= self.buffer_size:
                self._bulk_send_items(batch)
                items = []

        # If there are any items left, send them in a bulk:
        if len(batch) > 0:
            self._bulk_send_items(batch)

    def _send_items(self, batch):
        self.loaded_items_count += len(batch)
        logger.debug('Bulk writing {} items'.format(len(batch)))
        helpers.bulk(client=self.es, actions=batch)
```

In the context of the loop we can simply check if some items are left after the loop with `if len(items) > 0`, while this is much more complex to do in the context of the one item processing (or inside the loop).

We also don't need to keep the items buffer inside `self`, it's OK to pass variables across class methods.

I usually keep to this pattern: everything related to the object state - variables attached to `self` - should be defined in the constructor, I am not doing something like `self.new_field = ...` inside the class methods.